### DISCUSSION TITLE
Pagination with startkey, endkey and limit

### DISCUSSION THREAD

**Discussion by mardonedias** [apache/couchdb#3007] [Q&A] [answered]
Hello,

I have a little problem with couchdb, it's about pagination with startkey, endkey and limit. I can page to the next pages. But, I cannot understand how to keep the previous key to return. I use nodejs + express. If I could pass on some material with an example I would appreciate it. I've looked for a lot, a lot, but, there is nothing but the algorithm of the documentation itself. If you can help me I would appreciate it very much.

Thank you.

**Comment by wohali** [comment, marked as answer]
The generally recommended approach is to store the previous page's startkey/endkey/limit values for paging backwards.

**Reply by mardonedias** [reply]
In cookie?

**Reply by wohali** [reply]
Let's look at how Fauxton does it for the "All Documents" view as I page through it:

```
http://localhost:5984/skimdb_registry/_all_docs?skip=0&limit=21
http://localhost:5984/skimdb_registry/_all_docs?skip=20&limit=21
http://localhost:5984/skimdb_registry/_all_docs?skip=40&limit=21
```

So it's doing it purely via the # of documents to skip, and using a limit too. The challenge of course is that the "window" may change if the results being given by _all_docs changes, too.

If you use a start_key and a limit that is +1 of your actual limit, your next query can use that last result as the new start_key. To page backwards, you'd have to keep an array of all of the previous start_keys as well.

I'm not a web developer so I can't tell you the best way to store this in your application - maybe `localStorage`?

**Reply by mardonedias** [reply]
I had already thought about it, about storing, but, on the server. Which would not be very practical. But, with his explanation, I understood that the best strategy is to store in the client.

Thank you very much.
